### ISSUE TITLE
Meta issue: editor integrations

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#271]
Creating a parent issue to track and document the state of ruff's editor integrations.

**Comment by charliermarsh** [comment]
We should support and document integration with VS Code and PyCharm at the very least. I know there was some prior work + discussion in #118.

**Comment by charliermarsh** [comment]
Part of the work here could be making @harupy's VS Code extension publicly available. I'm happy to take that on if needed.

**Comment by Seamooo** [comment]
I've been playing around with the microsoft lsp as a way of making ruff available to a wider number of editors (particularly vscode and coc-nvim). The major blocker for me is that the main function for linting `ruff::linter::lint_path` only accepts a `Path` for the source. There's currently no public interface for linting from contents, or lower level access to the checking functions (i.e. `check_path` is private and `check_ast` requires `autofix::fixer::Mode` but `autofix` is private)

**Comment by charliermarsh** [comment]
I haven’t had time to dig in on this, but is the solution to exposure lint_path and then have the LSP use the ruff crate as a library? Or do we need to expose that functionality as a binary / CLI? I’m happy to do whichever’s necessary.

**Comment by charliermarsh** [comment]
In other words: if I added a lint_path function that was publicly exposed (with public args etc.), would that unblock this?

**Comment by charliermarsh** [comment]
Ok cool, very doable. Will put something together.

**Comment by charliermarsh** [comment]
(I added some documentation for PyCharm.)

**Comment by vikigenius** [comment]
@charliermarsh

Comenting on your question from https://github.com/charliermarsh/ruff/pull/289#issuecomment-1263866024 here since it's a more relevant thread.


> What does it take to enable others to use the LSP from VS Code and elsewhere?

Alternatively most python language servers are expected to be able to installed as a python package itself. So just distributing the language server as a python package using maturin would be enough.

Once that is done, the editors that support lsp would start developing clients to interact with the ruff language server.

Although personally I would like to have a general purpose language server to enable completions and use ruff as just an additional checker either as a separate binary or as an extension. Because the existing language servers for python are implemented in python and extremely slow or implemented in JS/TS (pyright) which I don't like.

Now that you have shown what we can do with Rust for python ecosystem hopefully in the future we can build an entire language server from the ground up in Rust for Python and have ruff as a checker extension for it. Like how Rust Analyzer is for Rust.

**Comment by charliermarsh** [comment]
👍 This makes sense! I want to spend some time exploring this soon (maybe this coming week), and putting together some official extensions. I use PyCharm, not VS Code, so I’m less familiar with the ecosystem (I need to understand the relationship between the LSP implementation, Ruff as a library, and then the actual editor extensions or actions and how those are distributed). But I do want Ruff to have the best possible editor support for VS Code given its popularity :)

**Comment by vikigenius** [comment]
Yep. As long as you have good language server support following the protocol. The people working on the editors will automatically start working on a client if the server is good/popular enough.

For instance, the moment we have a language server for ruff, I will try to contribute a client to Emacs https://github.com/emacs-lsp/lsp-mode so that I will be able to use it in Emacs. Similar would happen to all editors I am sure.

So the question basically becomes should the language server be a separate implementation? Or should it be somehow integrated into the same Repo (different crate) like how Rust-Analyzer does it for example. 

I know @Seamooo had an inital prototype implementation here https://github.com/Seamooo/ruffd it is a good POC for what a language server will look like in general.

**Comment by vikigenius** [comment]
The LSP specification https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/ is actually very readable unlike most specifications if you want a reference

Additionally something to consider is that all language servers will have different capabilities that they supports. A hypothetical `ruff-language-server` will support formatting related capabilities such as `codeActionProvider` and `diagnosticsProvider`.

But it won't support static checking and completion capabilities like `completionProvider` `declarationProvider` (goto declaration) etc.

A language server like Rust-Analyzer basically supports everything since it integrates, formatting, diagnostics, completion etc. all in one.

I know that linting itself is a huge task that you are taking on, but it might be useful if things like the parsing, AST/CST logic etc. can be well defined and reusable enough that in the future we can think of adding completion and type checking related capabilities as well for an all in one rust based python language server.

**Comment by charliermarsh** [comment]
Awesome. I'd like it to be in the same repo, if possible. We can use @Seamooo's implementation as a starting point (or would it ever make sense to build atop something like [`tower-lsp`](https://github.com/ebkalderon/tower-lsp)? I have to do some research here).

You're exactly right that it'd be good to design this with future capabilities in mind. Ruff already supports auto-fix, but auto-formatting is another good example of a capability that would be interesting to explore within the scope of the project.

**Comment by charliermarsh** [comment]
(Let's keep using this issue to discuss for now, and I'll open something else specific to the LSP if / when I have work ready for review / feedback.)

**Comment by vikigenius** [comment]
Oh, I didn't know about tower-lsp before. It seems like a pretty useful abstraction. The protocal itself is not that hard to implement as can be seen from the prototype implementation https://github.com/Seamooo/ruffd

But having something like tower-lsp will definitely make it easier and maybe more easily maintainable (hopefully).

**Comment by beatslikeahelix** [comment]
Are there any plans for neovim integration?

**Comment by Seamooo** [comment]
Definitely like the look of tower-lsp, and would be a big step for maintianability. Only reservervations would be having to dig into how it does scheduling as request ordering is important in the case of mutations.
If we're looking to create a code action provider for fixes, the api exposing Check, rather than Message, and making Fix a public type should enable everything.

**Comment by Seamooo** [comment]
@beatslikeahelix https://github.com/Seamooo/ruffd works with coc.nvim admittedly there isn't a release candidate right now as it's  very POC. To get it to work you can add below to your `coc-settings.json`
```json
"ruffd": {
	"command": "/path/to/ruffd",
	"filetypes": ["python"],
}
```

**Comment by charliermarsh** [comment]
I'm going to try building a VS Code extension atop `ruffd`.

**Comment by fannheyward** [comment]
coc-pyright has added ruff support, only lint for now.

**Comment by charliermarsh** [comment]
@fannheyward - That's awesome! Thank you for doing that. I'll plug it in the Ruff docs.

**Comment by JafarAbdi** [comment]
For those who are using neovim, you could integrate ruff directly with [efm](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#efm) just add the [following lines](https://github.com/JafarAbdi/myconfigs/blob/6f0b6b2450e92ec8fc50422928cd22005b919110/efm-langserver/config.yaml#L14-L20) to your config file

**Comment by fannheyward** [comment]
`coc-pyright` has added `ruff` fix support.

**Comment by charliermarsh** [comment]
Thanks @fannheyward!

**Comment by koxudaxi** [comment]
@charliermarsh 
I have started to develop the Pycharm plugin `ruff`
I will come back with more information.
https://github.com/koxudaxi/ruff-pycharm-plugin

**Comment by koxudaxi** [comment]
I implemented the three features on the [ruff pycharm plugin](https://github.com/koxudaxi/ruff-pycharm-plugin). 
- Run `ruff --fix` as an action
-  Integrate Reformat Code with `ruff --fix` 
-  Run `ruff --fix` for a file when the file is saved


And, We can install the plugin on PyCharm settings from [JetBrains marketplace](https://plugins.jetbrains.com/plugin/20574-ruff)
The version invokes `ruff` command and give the python code to `stdin`.

@charliermarsh 
As the next step, I would try to support `ruffd` (LSP).
I can't find the current status of the `ruffd`

Do you have the roadmap for supporting LSP?
If we need the time. I will skip the task now. And I will take another task.

**Comment by charliermarsh** [comment]
@koxudaxi - Awesome! I'll give it a try today.

I'm still figuring out the roadmap for `ruffd`. Right now, I'm slightly more interested in using the LSP from https://github.com/charliermarsh/vscode-ruff and seeing if I can leverage it outside of VS Code (e.g., in Sublime Text).

**Comment by charliermarsh** [comment]
The most relevant stuff in `vscode-ruff` is [here](https://github.com/charliermarsh/vscode-ruff/blob/main/bundled/tool/server.py) -- the bundled Python server running on `pygls`. You can see the supported actions in the `@LSP_SERVER.feature` decorators.

**Comment by charliermarsh** [comment]
(I know you asked about `ruffd`, but right now `vscode-ruff` is more familiar to me.)

**Comment by charliermarsh** [comment]
@koxudaxi - Does the extension surface errors in the editor? Or is it meant to be invoked with the "Run Ruff" action?

**Comment by koxudaxi** [comment]
This version runs `ruff --fix,` and it doesn't show any message, including an error from the command. 
I will expect to show the message in the `message window` like `console window` or/and highlight the code with the error message on the editor in future versions.
If ruff can inspect the code quickly, we can run the ruff by every user action(changing code).
Or, Can ruff partial inspection with cache?

**Comment by fannheyward** [comment]
> Run `ruff --fix` for a file when the file is saved

I did `ruff` linting in coc-pyright, also provided `fix` code action to pick and use.

But for `ruff --fix` on save, I don't have a strong desire to add. Apply fixes on save may change the whole file that bring unpredictable changes.

This is my OWN opinion, what do you think for this feature?

**Comment by charliermarsh** [comment]
My general take is that fix-on-save should be entirely opt-in and disabled by default.

**Comment by koxudaxi** [comment]
I agree with you. The feature is disabled by default.

**Comment by charliermarsh** [comment]
> If ruff can inspect the code quickly, we can run the ruff by every user action(changing code).

The Ruff VS Code plugin just runs constantly on the open file, I think it's fast enough to do it on-user-change.

**Comment by tsugumi-sys** [comment]
Are there any plans for [helix-editor](https://github.com/helix-editor/helix) integration?

**Comment by cole-jacobs** [comment]
> For instance, the moment we have a language server for ruff, I will try to contribute a client to Emacs https://github.com/emacs-lsp/lsp-mode so that I will be able to use it in Emacs. Similar would happen to all editors I am sure.

Any progress on this, @vikigenius ? Anything special I need to do to get this working in emacs?

**Comment by koxudaxi** [comment]
@charliermarsh 
I have published version v0.0.3 of the ruff pycharm plugin. 
I added the feature `inspection` and `quick fix`
Could you please try it? you can download it [marketplace](https://plugins.jetbrains.com/plugin/20574-ruff)
If you feel good, I want to put the link on the ruff documents. 
![quickfix](https://user-images.githubusercontent.com/630670/209247492-d79fdbb3-2f52-4ccd-af5f-0c60a9c38ab9.png)
![inspection](https://user-images.githubusercontent.com/630670/209247496-f1cfe9a8-6897-47ab-a518-0d334724f62e.png)


https://github.com/koxudaxi/ruff-pycharm-plugin

**Comment by charliermarsh** [comment]
@tsugumi-sys - I bet that https://github.com/charliermarsh/ruff-lsp would work with Helix, it's just a matter of configuring it.

Same for Emacs @cole-jacobs.

I can try to figure out how to integrate with those editors and add instructions to the `ruff-lsp` docs -- there are already a few examples, for Neovim and Sublime Text -- but would also love PRs to that effect.

**Comment by charliermarsh** [comment]
Done, thanks @koxudaxi!

**Comment by charliermarsh** [comment]
I'm going to close this for now since we have a good thing going with the LSP, VS Code extension, and so on. I'll file some issues on the LSP repo to figure out integrations with Emacs and Helix.

**Comment by fannheyward** [comment]
coc.nvim + ruff-lsp https://github.com/neoclide/coc.nvim/wiki/Language-servers#using-ruff-lsp

```json
"languageserver": {
  "ruff-lsp": {
    "command": "ruff-lsp",
    "filetypes": ["python"]
  }
}
```

**Comment by charliermarsh** [comment]
Amazing, thanks @fannheyward, I'll add those to the docs.

**Comment by charliermarsh** [comment]
Added Helix docs in https://github.com/charliermarsh/ruff-lsp/pull/20 @tsugumi-sys.

**Comment by tsugumi-sys** [comment]
@charliermarsh 
I can use ruff in my project now and Omg, ruff is so first and I love it! Thank you very much:)

**Comment by charliermarsh** [comment]
Yay, thanks so much for the kind words!

**Comment by kuator** [comment]
This is awesome!

**Comment by charliermarsh** [comment]
I spent a little bit of time installing `lsp-mode` with Emacs but couldn't figure out how to get `ruff-lsp` to work with it, just too unfamiliar with Emacs. If anyone is able to set it up, and wants to contribute instructions, it would be hugely appreciated.

See: https://github.com/charliermarsh/ruff-lsp/issues/19.

**Comment by charliermarsh** [comment]
@koxudaxi - Just wanted to say thanks again for building and maintaining the PyCharm plugin! I was playing with it today, and I love that you're surfacing the fix messages etc. in the UI.

**Comment by koxudaxi** [comment]
@charliermarsh 
Last week, I added ruff to my project https://github.com/koxudaxi/datamodel-code-generator
After using ruff in the actual project, I am excited about ruff and the PyCharm integration.
I will continue to maintain and add features to the PyCharm plugin.

Thanks for the feedback and your fantastic project.

PS. After knowing your commit history on ruff. I contribute to OSS every day 😸
